cloudigrade/util/aws.py

```diff
@@ -220,6 +229,120 @@ def get_ec2_instance(session, instance_id):
     return session.resource('ec2').Instance(instance_id)
 
 
+def get_ami(session, image_id, region):
+    """
+    Return an Amazon Machine Image running on an EC2 instance.
+
+    Args:
+        session (boto3.Session): A temporary session tied to a customer account
+        image_id (str): An AMI ID
+
+    Returns:
+        Image: A boto3 EC2 Image object.
+
+    """
+    return session.resource('ec2', region_name=region).Image(image_id)
+
+
+def get_ami_snapshot_id(ami):
+    """
+    Return the snapshot id from an Image object.
+
+    Args:
+        ami (boto3.resources.factory.ec2.Image): A machine image object
+
+    Returns:
+        string: The snapshot id for the machine image's root volume
+
+    """
+    for mapping in ami.block_device_mappings:
+        # For now we are focusing exclusively on the root device
+        if mapping['DeviceName'] != ami.root_device_name:
+            continue
+        return mapping.get('Ebs', {}).get('SnapshotId', '')
+
+
+def change_snapshot_ownership(session, snapshot_id, region, operation='add'):
+    """
+    Add or remove permissions to a snapshot.
+
+    Args:
+        session (boto3.Session): A temporary session tied to a customer account
+        snapshot_id (str): The id of the snapshot to modify
+        region (str): The region the machine image and snapshot reside in
+        operation (str): Should be one of 'add'|'remove'
+
+    Returns:
+        bool: Whether ownership has been modified
+
+    """
+    acceptable_operations = ('add', 'remove')
+    if operation not in acceptable_operations:
+        raise AwsValueError(_('{o} not one of {ao}').format(
+            o=operation,
+            ao=acceptable_operations)
+        )
+
+    snapshot = session.resource('ec2',
+                                region_name=region).Snapshot(snapshot_id)
+
+    if snapshot.encrypted:
+        logger.warning(_('Snapshot is encrypted. Not copying.'))
+        raise AwsSnapshotEncryptedError
+
+    attribute = 'createVolumePermission'
```

Comment:
Do we actually need `createVolumePermission`? It is my understanding that we only need to add our `userId` to the snapshot, not the `createVolumePermission` list too. Since we're copying the snapshot, not creating the volume from it. We create the volume from _our_ copied version.

Reply:
I'd love to use the minimum API call that works here. The best I can tell we have to use modify_attribute, and that has two options for attribute. either `productCodes` or `createVolumePermission` and we've been encountering funky validation with this specific API call recently. As written, it was the only way I could get the snapshot to be modified in a way that we could copy it.

If I'm missing something, please add here.

Reply:
@adberglund you're right, confusion on my part with the docs.